sdk/nodejs/resource.ts

```diff
@@ -987,6 +994,60 @@ export interface CustomResourceOptions extends ResourceOptions {
     // that mergeOptions works properly for it.
 }
 
+export class ResourceHook {
+    public name: string;
+    public handler: ResourceHookFunction;
+    public onDryRun: boolean;
+
+    /**
+     * Tracks the registration of the resource hook. The promise will resolve
+     * once the hook has been registered, or reject if any error occurs.
+     *
+     * @internal
+     */
+    public __registered: Promise<void>;
+
+    /**
+     * A private field to help with RTTI that works in SxS scenarios.
+     *
+     * @internal
+     */
+    public readonly __pulumiResourceHook: boolean = true;
+
+    constructor(name: string, handler: ResourceHookFunction, opts?: ResourceHookOptions) {
+        this.name = name;
+        this.handler = handler;
+        this.onDryRun = opts?.onDryRun || false;
+        this.__registered = registerResourceHook(this);
+    }
+
+    public static isInstance(obj: any): obj is ResourceHook {
+        return utils.isInstance<ResourceHook>(obj, "__pulumiResourceHook");
+    }
+}
+
+export interface ResourceHookOptions {
+    onDryRun?: boolean;
+}
+
+export interface ResourceHookArgs {
+    urn: URN;
+    id: ID;
+    inputs: Record<string, any>;
+    outputs: Record<string, any>;
+}
+
+export type ResourceHookFunction = (args: ResourceHookArgs) => void | Promise<void>;
+
+export interface ResourceHookBinding {
```

Comment:
Nit -- feels like a plural thing?
```suggestion
export interface ResourceHookBindings {
```